---
title: "Fig2H_FFPE-A"
author: "Javier Escudero"
date: "2023-11-30"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Loading libraries
```{r, message=FALSE}
rm(list = ls())

library(semla)
library(infercnv)
library(SpatialInferCNV)
library(ggplot2)
library(viridis)
library(Polychrome)
library(patchwork)
library(gtools)
library(ape)
library(phylogram)
library(gridExtra)
library(RColorBrewer)
library(scales)
library(ggbio)
library(gprofiler2)
library(ggpubr)
library(tidyverse)
```

Setting up directories
```{r}
st.dir <- "/Users/javierescudero/Documents/work/SIMPlex-seq/data/visium/human_BC/R_objects"
cnv.dir <- "/Users/javierescudero/Documents/work/SIMPlex-seq/results/inferCNV/iCNV_outs/Visium/sc_ref/Rezas_manual-annot/clean_snREF"
patho.dir <- "/Users/javierescudero/Documents/work/SIMPlex-seq/data/visium/human_BC"
out.dir <- "~/Documents/work/SIMPlex-seq/results/inferCNV/analysis/cleaned_snREF-visClones"
```

Some global variables
```{r}
save_plots <- TRUE
```

# Summary
Code required to recreate the spatial plot showed in Figure 2.G

# Analysis
## Load Visium data
FFPE slide
```{r}
se.ffpe <-  readRDS(list.files(st.dir, pattern = "ffpe_2", full.names = TRUE))
DefaultAssay(se.ffpe) <- "Spatial"

se.ffpe@tools[["Staffli"]]@imgs <- str_subset(list.files("/Users/javierescudero/Documents/work/SIMPlex-seq/data/visium/human_BC/FFPE", pattern = "hires", full.names = TRUE, recursive = TRUE), pattern = paste(unique(se.ffpe$section.name), collapse = "|"))
se.ffpe <- LoadImages(se.ffpe, image_height = 2000)
```

Loading annotation for FFPE slide
```{r}
ffpe.a <- read_csv(list.files(patho.dir, 
                              full.names = TRUE, pattern = "V42L22-105_A1_Epithelial", recursive = TRUE)) %>%
  drop_na(Epithelial)
colnames(ffpe.a) <- c("barcode", "anno")
ffpe.a <- ffpe.a %>% column_to_rownames(var = "barcode") 

se.ffpe <- AddMetaData(se.ffpe, metadata = ffpe.a)
```

Define a function for loading inferCNV outs
```{r}
load_icnv_outs <- function(cnv_dir, sample_interest, file_to_look){
  #Getting files from sample to focus on
  sample_dir <- str_subset(list.dirs(cnv_dir, full.names = TRUE), sample_interest)  
  sample_file <- list.files(sample_dir, full.names = TRUE, pattern = file_to_look)
  #Loading file
  if (str_detect(file_to_look, "dendrogram")){
    loaded_file <- read.tree(file = sample_file)
  } else {
    loaded_file <- readRDS(file = sample_file)
  }
  
  return(loaded_file)
}
```

## Exploring V42L22-105_A1
### Exploring dendrogram clusters
Generating dendrogram for section
```{r, eval=FALSE}
dendrogram <- as.hclust(load_icnv_outs(cnv_dir = cnv.dir,
                                       sample_interest = "V42L22-105_A1",
                                       file_to_look = "infercnv.observations_dendrogram.txt"))
```

In this section it doesn't make that much sense to divide into 5 clusters. Looking at the heatmap, the whole sections looks to be very uniform, only the spots close to those annotated as *Normal* seem to be a bit different. Based on the above classification, the spots that are different fall under cluster 4. So we can rephrase the above clusters to: **if you are not in 4, you belong to the same cluster. If you are in 4, you are indeed distinct**
```{r, eval=FALSE}
spots.dendro <- data.frame(cutree(dendrogram, k = 5))
colnames(spots.dendro) <- "icnv_clones"

spots.dendro <- spots.dendro %>% 
  mutate(icnv_clones = ifelse(icnv_clones == "4", "B", "A"))
```

Loading dendrogram
```{r}
dendrogram <- as.phylo(load_icnv_outs(cnv_dir = cnv.dir,
                                       sample_interest = "V42L22-105_A1",
                                       file_to_look = "infercnv.observations_dendrogram.txt"))
```

Manual selection of clones based on the dendrogram
```{r}
subtrees <- subtrees(dendrogram)
png(paste0(cnv.dir,"/iCNV_V42L22-105_A1/phylo.png"),width=10000,height=2500, res = 300)
plot(dendrogram,show.tip.label = FALSE)
nodelabels(text=1:dendrogram$Nnode,node=1:dendrogram$Nnode+Ntip(dendrogram))
dev.off()
```

Assigning spots to a icnv_clones manually depending on where they fall in the iCNV dendrogram. The clone calling I want to do is the following:
  1. **Clone A:** spots under node 2807
  2. **Clone B:** everything else
```{r}
nodes.interest <- c(2807,2831, 2926, 2)
spots.dendro <- map_dfr(nodes.interest, function(node){
  SelectingSubTreeData(subtrees, node)
})
```

Formatting and adding to Seurat
```{r}
spots.dendro <- spots.dendro %>% 
  column_to_rownames(var = "Barcode") %>% 
  mutate(Node = ifelse(Node == "Node_2807", "A", 
                       ifelse(str_detect(Node, "2831|2926|2"), "B","hm")))
colnames(spots.dendro) <- "icnv_clones"
```

Adding dendrogram information to Seurat object
```{r}
se.sub <- SubsetSTData(se.ffpe, spots = rownames(spots.dendro))
se.sub <- AddMetaData(se.sub, spots.dendro)
```

Spatial visualization of iCNV-based clusters
```{r}
plot <- MapLabels(se.sub , column_name = "icnv_clones", image_use = "raw",
            pt_size = 1.5, override_plot_dims = FALSE, label_by = "section.name",
            colors = viridis::viridis(n = length(unique(se.sub$icnv_clones))),
            # colors =  RColorBrewer::brewer.pal(length(unique(se.sub$icnv_clones)), "Set1"),
            pt_alpha = 1) &
    guides(fill = guide_legend(override.aes = list(size = 9))) &
    theme(legend.position = "right")

plot

plot.dir <- paste0(c(out.dir, "clone_spots"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir, recursive = TRUE)
}

if(save_plots == TRUE) {
  ggsave(plot, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clone_spots.pdf"),
         width = 8, height = 8, units = "in")
} else {
}
```
We can visualize the pathology annotation and the iCNV clusters
```{r}
plot_df <- se.sub[[]] %>% 
  select(c(anno, icnv_clones)) %>% 
  group_by(icnv_clones, anno) %>% 
  summarise(n = n()) %>% 
  mutate(freq = n/sum(n))

plot <- ggplot(plot_df, aes(x = icnv_clones, y = freq, fill = anno)) +
  geom_bar(position = "fill", stat = "identity") +
  geom_text(aes(label = n),
            colour = "black", fontface = "bold",
            size = 3,
            position = position_stack(vjust = 0.5)) +
  # position = position_dodge(0.9)) +
  scale_fill_manual(values = viridis::turbo(n = 5)) +
  theme_light() +
  xlab("inferCNV clone") +
  ylab("prop. spots") +
  ggtitle("Presence of inferCNV clones in each pathologist-annotated region") +
  theme(plot.title = element_text(face="bold"))

plot

plot.dir <- paste0(c(out.dir, "clone_patho"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir, recursive = TRUE)
}

if(save_plots == TRUE) {
  ggsave(plot, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clone_patho.pdf"),
         width = 8, height = 4, units = "in")
} else {
}
```
### Spatial plotting of iCNV scores
First I will load the inferCNV scores for our section
```{r}
st.cnv <- load_icnv_outs(cnv_dir = cnv.dir,
                         sample_interest = "V42L22-105_A1",
                         file_to_look = "run.final.infercnv_obj")
```

Then I create the annotation
```{r}
cytoband <- read.delim("~/Documents/work/useful_stuff/cytoBandIdeo.txt", header = FALSE,
                       col.names = c("chr","chromStart","chromEnd","name","gieStain"))
```

I am going to define a gene grouping based on a window of 100 genes and on the chromosome arm. So each chromosome arm will be divided in lists of 100 genes.
```{r}
#we define the coordinates of each chromosome arm
chr.anno <- cytoband %>%
  mutate(arm = str_extract(name, "p|q")) %>%
  filter(!str_detect(chr, "_")) %>% #removing weird chromosome names
  filter(!str_detect(chr, "M")) %>% #we do not need the mitochondrial chromosome
  group_by(chr, arm) %>%
  summarise(max = max(chromEnd)) %>%
  pivot_wider(names_from = arm, values_from = max)

#now let's define the arms in which each gene is located, based on their starts coordinates
genes.anno <- rownames_to_column(st.cnv@gene_order, var = "gene_id") %>%
  inner_join(chr.anno, by="chr") %>%
  mutate(gene_arm = ifelse(start < p, "p", "q")) %>%
  unite("chrom", c('chr', 'gene_arm'), sep = "_", remove = FALSE) %>%
  group_by(chr) %>%
  mutate(window = paste0(chr, "_w", (row_number()-1) %/% 100 + 1)) %>%
  ungroup() %>%
  select(c(gene_id, start, stop,chr, chrom, window))
```

I am going to define a gene grouping based on a window of 100 genes and on the chromosome arm. So each chromosome arm will be divided in lists of 100 genes.
```{r}
#now let's define the arms in which each gene is located, based on their starts coordinates
genes.anno <- rownames_to_column(st.cnv@gene_order, var = "gene_id") %>%
  inner_join(chr.anno, by="chr") %>%
  mutate(gene_arm = ifelse(start < p, "p", "q")) %>%
  unite("chrom", c('chr', 'gene_arm'), sep = "_", remove = FALSE) %>%
  group_by(chr) %>%
  #group_by(chrom) %>%
  mutate(window = paste0(chr, "_w", (row_number()-1) %/% 100 + 1)) %>%
  ungroup() %>%
  select(c(gene_id, start, stop,chr, chrom, window))
```

I am computing the average CNV score in each spot for a specific genomic window. The metric I'm using right now is the mean of the window. If the window has only 1 gene, we skip that window.
```{r}
spot.cnv <- st.cnv@expr.data

names <- genes.anno %>% group_by(window) %>%
  filter(n() > 1) %>%
  tally() %>% pull(window)
gene.list <- genes.anno %>%
  group_by(window) %>% 
  filter(n() > 1) %>%
  group_map(~ pull(.x, gene_id)) %>%
  setNames(names)

# computing region-wise cnv averages. In this case, compute the average cnv score for each chromosome arm
spatial.cnv.win <- map(gene.list, function(region){
  cnv_arm <- colMeans(spot.cnv[rownames(spot.cnv) %in% region,], na.rm = TRUE)
}) %>% setNames(names(gene.list))
```

Turn the list of genomic regions into df
```{r}
spatial.cnv.df<- imap_dfc(spatial.cnv.win, function(x,id){
  df <- data.frame(id = x) 
  colnames(df) <- id
  return(df)
})
```

#### Looking at specific genes
It can be of interest to look at the predicted CNV score of genes of clinical interest in Breast Cancer. We can look at the estrogen receptor (*ESR1*), progesterone receptor (*PGR*) and human epidermal growth factor receptor 2 (*ERBB2*).

Quickly check which genes made it into the inferCNV analysis:
```{r}
genes.interest <- c("ESR1","PGR","ERBB2")
t.genes.interest <- genes.interest[genes.interest %in% rownames(spot.cnv)]
```

Retrieving the iCNV scores of these genes and adding to the Seurat for plotting
```{r}
gene.cnv <- as.data.frame(t(spot.cnv[t.genes.interest, colnames(se.sub)]))

se.sub <- AddMetaData(se.sub, gene.cnv)
```

##### Spatial plots
It might be interesting to also **plot the chromosomal window where the gene of interest is located**.
```{r}
# Getting the regions of the genes of interest
regions <- genes.anno %>% 
  filter(gene_id %in% t.genes.interest) %>% 
  mutate(sup_window = paste(gene_id, window, sep = "_")) 
# Retrieving the CNV scores of those regions, as maybe they were not selected as the most sd regions
regions.df <- spatial.cnv.df %>%
  select(all_of(regions$window))
colnames(regions.df) <- regions$sup_window
# Adding to seurat
se.sub <- AddMetaData(se.sub, regions.df)
```

To ensure that outliners of the data do not dominate the color scale of the plotting, I am going to define some limits following guidelines establish in the package inferCNV and used for colouring the inferCNV heatmap, and establishing 1 as the center of the color scale.
```{r}
regions.mtx <- as.matrix(spatial.cnv.df %>%
                                  select(all_of(regions$window)))
quantiles <- quantile(regions.mtx, c(0.01,0.99))

#establishing max and min thresholds considering the maximum distance of any cnv score from the center 1
spread <- max(abs(c(1 - quantiles[1], 1 - quantiles[2])))
min.thresh <- round(1 - spread, 1)
max.thresh <- round(1 + spread, 1)

#transforming data matrix to remove outliners
regions.mtx[regions.mtx < min.thresh] <- min.thresh
regions.mtx[regions.mtx > max.thresh] <- max.thresh
```

Plotting clinical genes of interest
```{r}
p1 <- MapFeatures(se.sub, features = t.genes.interest, image_use = "raw", override_plot_dism = TRUE,
                  pt_size = 0.75
                  ) +
  plot_layout(guides = "collect") &
  scale_fill_gradientn(colours = rev(RColorBrewer::brewer.pal(n = 11, name = "RdBu")), 
                       values = rescale(c(min.thresh, 1, max.thresh)),
                       guide = "colorbar", limits=c(min.thresh,max.thresh),
                       oob = squish) &
  theme(plot.title = element_blank(),
        legend.position = "right", legend.text = element_text(angle = 0),
        legend.key.size = unit(1.5, 'cm'),
        text = element_text(size = 20)) &
  labs(fill = "iCNV_score")

p1

plot.dir <- paste0(c(out.dir, "spatial_iCNVs"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir)
}

if(save_plots == TRUE) {
  ggsave(p1, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clinical_genes.pdf"),
         width = 9, height = 7, units = "in")
} else {
}
```

Plotting windows of genes of interest
```{r}
p1 <- MapFeatures(se.sub, features = colnames(regions.df), image_use = "raw", override_plot_dism = TRUE,
                  pt_size = 0.75
                  ) +
  plot_layout(guides = "collect") &
  scale_fill_gradientn(colours = rev(RColorBrewer::brewer.pal(n = 11, name = "RdBu")), 
                       values = rescale(c(min.thresh, 1, max.thresh)),
                       guide = "colorbar", limits=c(min.thresh,max.thresh),
                       oob = squish) &
  theme(plot.title = element_blank(),
        legend.position = "right", legend.text = element_text(angle = 0),
        legend.key.size = unit(1.5, 'cm'),
        text = element_text(size = 20)) &
  labs(fill = "iCNV_score")

p1

plot.dir <- paste0(c(out.dir, "spatial_iCNVs"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir)
}

if(save_plots == TRUE) {
  ggsave(p1, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clinical_windows.pdf"),
         width = 9, height = 7, units = "in")
} else {
}
```
##### Violin plots
It is interesting to see the scores of each gene in each spot by annotation.
```{r}
feat <- c(t.genes.interest, "anno")
plot_df <- se.sub[[]] %>% 
  select(all_of(feat)) %>% 
  rownames_to_column(var = "barcode") %>% 
  pivot_longer(cols = !c(anno, barcode), names_to = "gene", values_to = "cnv_score")

p <- ggplot(plot_df, aes(x = gene, y = cnv_score,
                                fill = gene)) +
  geom_violin(scale = "count") +
  geom_boxplot(width=0.1, color = "darkgreen") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "black") +
  # geom_jitter(size = 0.5)  +
  facet_grid(~anno, scales = "free", space = "free") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),text = element_text(size=25)) +
  theme_light() +
  xlab("clinically relevant genes") +
  ylab("cnv score") +
  ggtitle("Predicted CNV score on spots annotated by pathologist") +
  theme(plot.title = element_text(face="bold"))
p

plot.dir <- paste0(c(out.dir, "clinicalGenes_by_pathology"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir)
}

if(save_plots == TRUE) {
  ggsave(p, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clinical_genes.pdf"),
         width = 9, height = 7, units = "in")
} else {
}
```

It is interesting to see the scores of the chromosomal windows where the genes of interest are located in each spot by annotation.
```{r}
feat <- c(colnames(regions.df), "anno")
plot_df <- se.sub[[]] %>% 
  select(all_of(feat)) %>% 
  rownames_to_column(var = "barcode") %>% 
  pivot_longer(cols = !c(anno, barcode), names_to = "gene_window", values_to = "cnv_score")

p <- ggplot(plot_df, aes(x = gene_window, y = cnv_score,
                         fill = gene_window)) +
  geom_violin(scale = "count") +
  geom_boxplot(width=0.1, color = "darkgreen") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "black") +
  # geom_jitter(size = 0.5)  +
  facet_grid(~anno, scales = "free", space = "free") +
  theme_light() +
  xlab("clinically relevant genes") +
  ylab("cnv score") +
  ggtitle("Predicted CNV score on spots annotated by pathologist") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(face="bold"))

p

plot.dir <- paste0(c(out.dir, "clinicalGenes_by_pathology"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir)
}

if(save_plots == TRUE) {
  ggsave(p, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clinical_windows.pdf"),
         width = 9, height = 7, units = "in")
} else {
}
```

It is interesting to see the scores of each gene in each spot by iCNV clone
```{r}
feat <- c(t.genes.interest, "icnv_clones")
plot_df <- se.sub[[]] %>% 
  select(all_of(feat)) %>% 
  rownames_to_column(var = "barcode") %>% 
  pivot_longer(cols = !c(icnv_clones, barcode), names_to = "gene", values_to = "cnv_score")

p <- ggplot(plot_df, aes(x = gene, y = cnv_score,
                                fill = gene)) +
  geom_violin(scale = "count") +
  geom_boxplot(width=0.1, color = "darkgreen") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "black") +
  # geom_jitter(size = 0.5)  +
  facet_grid(~icnv_clones, scales = "free", space = "free") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),text = element_text(size=25)) +
  theme_light() +
  xlab("clinically relevant genes") +
  ylab("cnv score") +
  ggtitle("Predicted CNV score on spots belonging to inferred clones") +
  theme(plot.title = element_text(face="bold"))
p

plot.dir <- paste0(c(out.dir, "clinicalGenes_by_clone"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir)
}

if(save_plots == TRUE) {
  ggsave(p, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clinical_genes.pdf"),
         width = 9, height = 7, units = "in")
} else {
}
```

It is interesting to see the scores of the chromosomal windows where the genes of interest are located in each spot by icnv_clones
```{r}
feat <- c(colnames(regions.df), "icnv_clones")
plot_df <- se.sub[[]] %>% 
  select(all_of(feat)) %>% 
  rownames_to_column(var = "barcode") %>% 
  pivot_longer(cols = !c(icnv_clones, barcode), names_to = "gene_window", values_to = "cnv_score")

p <- ggplot(plot_df, aes(x = gene_window, y = cnv_score,
                         fill = gene_window)) +
  geom_violin(scale = "count") +
  geom_boxplot(width=0.1, color = "darkgreen") +
  geom_hline(yintercept = 1, linetype = "dashed", color = "black") +
  # geom_jitter(size = 0.5)  +
  facet_grid(~icnv_clones, scales = "free", space = "free") +
  theme_light() +
  xlab("clinically relevant genes") +
  ylab("cnv score") +
  ggtitle("Predicted CNV score on spots belonging to inferred clones") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        plot.title = element_text(face="bold"))

p

plot.dir <- paste0(c(out.dir, "clinicalGenes_by_clone"), collapse = "/")
if(dir.exists(plot.dir)) {
} else {
  dir.create(plot.dir)
}

if(save_plots == TRUE) {
  ggsave(p, 
         filename = paste0(plot.dir, "/", unique(se.sub$section.name), "_clinical_windows.pdf"),
         width = 9, height = 7, units = "in")
} else {
}
```

# Metadata
```{r}
date()
sessionInfo()
```


